### PR TITLE
Add Grpc Protos used for the Vulkan Layer.

### PR DESCRIPTION
**Pull request by florian-kuebler**
This adds a CaptureEvent proto for a Vulkan submission, which gathers
information from a single submission (VkQueueSubmit). In particular
it contains the timings for all the command buffers in that submission
and the debug markers opend/closed in that submission.

These protos can be then used by the layer to send information to the
service.

Test: Compile

### CODE REVIEW

File: OrbitGrpcProtos/capture.proto [hunk 1]
```diff
@@ -191,5 +235,6 @@ message CaptureEvent {
     InternedTracepointInfo interned_tracepoint_info = 9;
     TracepointEvent tracepoint_event = 10;
     IntrospectionScope introspection_scope = 12;
+    GpuQueueSubmission gpu_queue_submission = 13;
```
Comment (dpallotti) [inline OrbitGrpcProtos/capture.proto:238]: I would suggest moving it right after `GpuJob`, it makes sense to me if they are next to each other.

File: OrbitGrpcProtos/capture.proto [hunk 2]
```diff
@@ -177,6 +177,50 @@ message AddressInfo {
...
+
+message GpuDebugMarkerBeginInfo {
+  GpuQueueSubmissionMetaInfo meta_info = 1;
+  uint64 gpu_timestamp_ns = 2;
+}
+
+message GpuDebugMarker {
+  oneof begin_marker_option {
+    GpuDebugMarkerBeginInfo begin_marker = 1;
+  };
```
Comment (dpallotti) [inline OrbitGrpcProtos/capture.proto:210]: Why `oneof` when you have only one option?

File: OrbitGrpcProtos/capture.proto [hunk 3]
```diff
@@ -177,6 +177,50 @@ message AddressInfo {
...
 
+message GpuQueueSubmission {
+  GpuQueueSubmissionMetaInfo meta_info = 1;
+  repeated GpuSubmitInfo submit_infos = 2;
+  repeated GpuDebugMarker completed_markers = 3;
+  uint32 num_begin_markers = 4;
+}
+
+message GpuQueueSubmissionMetaInfo {
+  int32 thread_id = 1;
```
Comment (dpallotti) [inline OrbitGrpcProtos/capture.proto:188]: All the other protos here call it `tid`, I would suggest consistency.

File: OrbitGrpcProtos/capture.proto [hunk 4]
```diff
@@ -177,6 +177,50 @@ message AddressInfo {
...
+  uint64 gpu_timestamp_ns = 2;
+}
+
+message GpuDebugMarker {
+  oneof begin_marker_option {
+    GpuDebugMarkerBeginInfo begin_marker = 1;
+  };
+  uint64 end_gpu_timestamp_ns = 2;
+  uint64 text_key = 3;
+  uint64 depth = 4;
```
Comment (dpallotti) [inline OrbitGrpcProtos/capture.proto:213]: This should probably be `int32`, similarly to `FunctionCall`, unless you have a specific reason why you need it unsigned (and 64 bits).

File: OrbitGrpcProtos/capture.proto [hunk 5]
```diff
@@ -177,6 +177,50 @@ message AddressInfo {
   }
 }
 
+message GpuQueueSubmission {
+  GpuQueueSubmissionMetaInfo meta_info = 1;
+  repeated GpuSubmitInfo submit_infos = 2;
+  repeated GpuDebugMarker completed_markers = 3;
+  uint32 num_begin_markers = 4;
```
Comment (dpallotti) [inline OrbitGrpcProtos/capture.proto:184]: This should probably be `int32` unless there's a specific reason for it to be unsigned.

File: OrbitGrpcProtos/capture.proto (hunk 2 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:210]: That is an purely optional field. Unfortunately, the version of ProtoBuf that we are using, does not support `optional`.

Comment (dpallotti) [reply OrbitGrpcProtos/capture.proto:210]: That's not the point of `oneof`. In `proto3` all fields are already optional.

Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:210]: But then I can't check if that field is there.
https://stackoverflow.com/questions/42622015/how-to-define-an-optional-field-in-protobuf-3/42634681

Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:210]: Wait, I was wrong. That was/is the case for primitives. At the beginning, that field should be a primitive. For the others, there are also the ` has` methods.

Comment (dpallotti) [reply OrbitGrpcProtos/capture.proto:210]: I have to give it to you that it's a bit of a mess for primitives, I forgot about that. But yes, in this case there should be a `has_begin_marker`.

Comment (dpallotti) [reply OrbitGrpcProtos/capture.proto:210]: Maybe if you want to make it clearer just from the name you could rename `begin_marker` to `optional_begin_marker`, but to me that's not necessary.

Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto]: Changed it accordninly.

File: OrbitGrpcProtos/capture.proto (hunk 3 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto]: Done.

File: OrbitGrpcProtos/capture.proto (hunk 1 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:236]: Done.

File: OrbitGrpcProtos/capture.proto (hunk 4 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto]: Done.

File: OrbitGrpcProtos/capture.proto (hunk 5 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto]: Done.

File: OrbitGrpcProtos/capture.proto (hunk 2 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto]: > Maybe if you want to make it clearer just from the name you could rename begin_marker to optional_begin_marker, but to me that's not necessary.

I think it is fine as it is.

File: OrbitGrpcProtos/capture.proto (hunk 1 above)
Comment (dpallotti) [reply OrbitGrpcProtos/capture.proto:236]: I meant also here in `message CaptureEvent` (even more now that the order of definition of the messages is like that).

**Review by dpallotti — APPROVED** [review]

File: OrbitGrpcProtos/capture.proto [hunk 6]
```diff
@@ -104,7 +104,7 @@ message InternedTracepointInfo {
 message TracepointEvent {
   int32 pid = 1;
   int32 tid = 2;
-  int64 time = 3;
+  uint64 time = 3;
```
Comment (ronaldfw) [inline OrbitGrpcProtos/capture.proto:107]: Why this change? Seems to be unrelated to the rest?

File: OrbitGrpcProtos/capture.proto [hunk 7]
```diff
@@ -128,6 +128,48 @@ message GpuJob {
   uint64 dma_fence_signaled_time_ns = 11;
 }
 
+message GpuQueueSubmission {
+  GpuQueueSubmissionMetaInfo meta_info = 1;
+  repeated GpuSubmitInfo submit_infos = 2;
+  repeated GpuDebugMarker completed_markers = 3;
+  int32 num_begin_markers = 4;
```
Comment (ronaldfw) [inline OrbitGrpcProtos/capture.proto:135]: Is this the total number of markers begun or the remaining open markers?

File: OrbitGrpcProtos/capture.proto [hunk 8]
```diff
@@ -128,6 +128,48 @@ message GpuJob {
...
+  repeated GpuCommandBuffer command_buffers = 1;
+}
+
+message GpuCommandBuffer {
+  uint64 begin_gpu_timestamp_ns = 1;
+  uint64 end_gpu_timestamp_ns = 2;
+}
+
+message GpuDebugMarkerBeginInfo {
+  GpuQueueSubmissionMetaInfo meta_info = 1;
```
Comment (ronaldfw) [inline OrbitGrpcProtos/capture.proto:154]: It doesn't seem to be obvious why this message has a GpuQueueSubmissionMetaInfo field. Is this only for convenience and has the same data as the GpuQueueSubmission parent or is this different?

File: OrbitGrpcProtos/capture.proto (hunk 7 above)
Comment (ronaldfw) [reply OrbitGrpcProtos/capture.proto:135]: Independent of the answer, consider adding a comment in the code to clarify.

File: OrbitGrpcProtos/capture.proto (hunk 8 above)
Comment (ronaldfw) [reply OrbitGrpcProtos/capture.proto:154]: Same here, add a comment in the code to clarify why this is needed.

File: OrbitGrpcProtos/capture.proto (hunk 6 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:107]: Hah, good point. I think that is a left over from extracting the change from my branch. I'd actually like to keep it in, as this might clearly be an issue, though.

File: OrbitGrpcProtos/capture.proto (hunk 7 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:135]: that is the total number of begin markers in that submission.

File: OrbitGrpcProtos/capture.proto (hunk 8 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:154]: That differs, if the BeginMarkers originate from a different submission.

File: OrbitGrpcProtos/capture.proto (hunk 6 above)
Comment (ronaldfw) [reply OrbitGrpcProtos/capture.proto:107]: I would not leave changes in a PR with a different purpose, in particular if they have the potential to break things.

File: OrbitGrpcProtos/capture.proto (hunk 8 above)
Comment (ronaldfw) [reply OrbitGrpcProtos/capture.proto:154]: Ok, that should be explained in a code comment.

File: OrbitGrpcProtos/capture.proto [hunk 9]
```diff
@@ -128,6 +128,48 @@ message GpuJob {
...
+message GpuDebugMarker {
+  GpuDebugMarkerBeginInfo begin_marker = 1;
+  uint64 end_gpu_timestamp_ns = 2;
+  uint64 text_key = 3;
+  int32 depth = 4;
+  Color color = 5;
+}
+
+message Color {
+  float red = 1;
```
Comment (ronaldfw) [inline OrbitGrpcProtos/capture.proto:167]: Clarify the range of the numbers in a comment. Probably [0.f, 1.f], but I've also seen [0.f, 255.f] being used (whether that is good practice is a different matter).

File: OrbitGrpcProtos/capture.proto (hunk 8 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:154]: Done.

File: OrbitGrpcProtos/capture.proto (hunk 6 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:107]: Okay, agree.

Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto]: Removed it.

File: OrbitGrpcProtos/capture.proto (hunk 9 above)
Comment (florian-kuebler) [reply OrbitGrpcProtos/capture.proto:170]: Done.

**Comment by florian-kuebler** [comment]
@ronaldfw is there anything else from your side to add?
